libs/datasets/timeseries.py

```diff
@@ -1224,3 +1228,99 @@ def aggregate_puerto_rico_from_counties(dataset: MultiRegionDataset) -> MultiReg
+class AnnotationField(GetByValueMixin, ValueAsStrMixin, FieldName, enum.Enum):
+    VARIABLE = PdFields.VARIABLE
+    LOCATION_ID = CommonFields.LOCATION_ID
+    DATE = CommonFields.DATE
+    TYPE = "annotation_type"
+    COMMENT = "comment"
+
+
+@enum.unique
+class AnnotationType(GetByValueMixin, ValueAsStrMixin, enum.Enum):
+    CUMULATIVE_TAIL_TRUNCATED = "cumulative_tail_truncated"
+    CUMULATIVE_LONG_TAIL_TRUNCATED = "cumulative_long_tail_truncated"
+
+
+@dataclass
+class TailFilter:
+    annotations: List[Mapping[str, Any]] = dataclasses.field(default_factory=list)
+
+    # Counts that track what the filter has done.
+    skipped_too_short: int = 0
+    skipped_na_mean: int = 0
+    all_good: int = 0
+    truncated: int = 0
+    long_truncated: int = 0
+
+    @staticmethod
+    def run(
+        dataset: MultiRegionDataset, fields: List[FieldName]
+    ) -> Tuple["TailFilter", MultiRegionDataset]:
+        """Returns a dataset with recent data that looks bad removed from fields."""
+        timeseries_wide_dates = dataset.timeseries_wide_dates()
+
+        fields_mask = timeseries_wide_dates.index.get_level_values(PdFields.VARIABLE).isin(fields)
+        to_filter = timeseries_wide_dates.loc[pd.IndexSlice[:, fields_mask], :]
+        not_filtered = timeseries_wide_dates.loc[pd.IndexSlice[:, ~fields_mask], :]
+
+        tail_filter = TailFilter()
+        filtered = to_filter.apply(tail_filter._filter_one_series, axis=1)
+
+        merged = pd.concat([not_filtered, filtered])
+        timeseries_wide_variables = merged.stack().unstack(PdFields.VARIABLE).sort_index()
+
+        return tail_filter, dataclasses.replace(dataset, timeseries=timeseries_wide_variables)
+
+    def annotations_as_dataframe(self):
+        return pd.DataFrame(self.annotations)
+
+    def _filter_one_series(self, series_in: pd.Series) -> pd.Series:
+        """Filters one timeseries of cumulative values. This is a method so self can be used to
+        store side outputs."""
+
+        if len(series_in) < 28:
+            self.skipped_too_short += 1
+            return series_in
+        diff = series_in.diff()
+        # Treat the mean value 2 - 4 weeks ago as good.
+        mean = diff[-28:-14].mean()
+        if pd.isna(mean):
+            self.skipped_na_mean += 1
+            return series_in
+        # Pick a bound used to determine if the diff of recent values are reasonable. For
+        # now values less than 100th of the mean are considered stalled. The variance could be
+        # used to create a tighter bound but the simple threshold seems to work for now.
+        # TODO(tom): Experiment with other ways to calculate the bound.
+        threshold = math.floor(mean / 100)
+        # Go backwards in time, starting with the most recent observation. Stop as soon as there
+        # is an observation with diff that looks reasonable / not stalled / over the threshold.
+        count_observation_diff_under_threshold = 0
+        for i in range(1, 15):
```

Comment:
is there a large benefit to going back this far? this seems like a long way back to be checking.  But also I guess by this threshold, any region with a case count avg of < 100 won't ever apply any checks which seems reasonable.

Reply:
I think one of the stalls that motivated this change was ~9 days long.